### PR TITLE
tlock: init at 1.0.0

### PR DESCRIPTION
**Pull request by eklairs**
## Description of changes

tlock - Two-Factor Authentication Tokens Manager in Terminal

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SigmaSquadron** [comment]
Hi @eklairs! Will you be maintaining any packages? Generally, `maintainer-list` entries are added as separate commits alongside a package adoption/update/fix.

**Comment by eklairs** [comment]
Hi @SigmaSquadron!
Oh i did not know about that - I will be maintaining a package, I thought I will make two PRs for it (one for maintainership and other for adoption). I will add the package adoption commit with this PR :)

File: pkgs/by-name/tl/tlock/package.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}:
+
+buildGoModule rec {
+  pname = "tlock";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "eklairs";
+    repo = pname;
+    rev = "refs/tags/v${version}";
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:14]:
```suggestion
    rev = "v${version}";
```
Just `v${version}` should be fine.

File: pkgs/by-name/tl/tlock/package.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  buildGoModule,
+}:
+
+buildGoModule rec {
+  pname = "tlock";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "eklairs";
+    repo = pname;
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:13]:
```suggestion
    repo = "tlock";
```
Altering `pname` should not change the final build result.

File: pkgs/by-name/tl/tlock/package.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    rev = "refs/tags/v${version}";
+    hash = "sha256-O6erxzanSO5BjMnSSmn89w9SA+xyHhp0SSDkCk5hp8Y=";
+    fetchSubmodules = true;
+  };
+
+  vendorHash = "sha256-G402CigSvloF/SI9Wbcts/So1impMUH5kroxDD/KKew=";
+
+  doCheck = false;
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:21]:
```suggestion
  doCheck = false;
```
Do add a comment explaining why the checks won't pass.

File: pkgs/by-name/tl/tlock/package.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  ];
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X github.com/eklairs/tlock/tlock-internal/constants.VERSION=v${version}"
+  ];
+
+  meta = with lib; {
+    mainProgram = pname;
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:35]:
```suggestion
    mainProgram = "tlock";
```

File: pkgs/by-name/tl/tlock/package.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    "scripts"
+  ];
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X github.com/eklairs/tlock/tlock-internal/constants.VERSION=v${version}"
+  ];
+
+  meta = with lib; {
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:34]:
```suggestion
  meta = {
```
And edit the other attributes appropriately. (e.g. `licenses` -> `lib.licenses`)

File: pkgs/by-name/tl/tlock/package.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    mainProgram = pname;
+    license = licenses.mit;
+    homepage = "https://github.com/eklairs/tlock";
+    description = "Two-Factor Authentication Tokens Manager in Terminal";
+    changelog = "https://github.com/eklairs/tlock/releases/tag/v${version}";
+    maintainers = with maintainers; [ eklairs ];
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:40]:
```suggestion
    maintainers = with maintainers; [ eklairs ];
    platforms = lib.platforms.PLATFORM
```
Where `PLATFORM` is `linux`, `darwin`, `unix`, or something more exotic. If it's just x86_64 Linux, for instance, use `platforms = [ "x86_64-linux" ];`.

File: pkgs/by-name/tl/tlock/package.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  ];
+
+  meta = {
+    mainProgram = "tlock";
+    license = lib.licenses.mit;
+    homepage = "https://github.com/eklairs/tlock";
+    description = "Two-Factor Authentication Tokens Manager in Terminal";
+    changelog = "https://github.com/eklairs/tlock/releases/tag/v${version}";
+    maintainers = with lib.maintainers; [ eklairs ];
+    platforms = lib.platforms.unix ++ lib.platforms.windows;
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:39]:
```suggestion
    platforms = lib.platforms.unix;
```

Nixpkgs does not build for Windows.

File: pkgs/by-name/tl/tlock/package.nix [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  excludedPackages = [
+    "bubbletea"
+    "scripts"
+  ];
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X github.com/eklairs/tlock/tlock-internal/constants.VERSION=v${version}"
+  ];
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:30]:
```suggestion
```
Are these really necessary?

File: pkgs/by-name/tl/tlock/package.nix [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    hash = "sha256-O6erxzanSO5BjMnSSmn89w9SA+xyHhp0SSDkCk5hp8Y=";
+    fetchSubmodules = true;
+  };
+
+  vendorHash = "sha256-G402CigSvloF/SI9Wbcts/So1impMUH5kroxDD/KKew=";
+
+  excludedPackages = [
+    "bubbletea"
+    "scripts"
+  ];
```
Comment (SigmaSquadron) [inline pkgs/by-name/tl/tlock/package.nix:24]:
```suggestion
```
Why aren't these being packaged?

File: pkgs/by-name/tl/tlock/package.nix (hunk 8 above)
Comment (eklairs) [reply pkgs/by-name/tl/tlock/package.nix:30]: Yes, they are.
1. `-s` and `-w` are used to reduce the size of the binary
2. `"-X github.com/eklairs/tlock/tlock-internal/constants.VERSION=v${version}"` is used to override the value of a constant during time with the version of the package (which is then used internally to show the package version)

File: pkgs/by-name/tl/tlock/package.nix (hunk 9 above)
Comment (eklairs) [reply pkgs/by-name/tl/tlock/package.nix:24]: 1. `scripts` - This folder contains bash scripts for development stuff so there is no need for it to be packaged.

**Review by SigmaSquadron — APPROVED** [review]
Thank you for answering my questions! This looks good to me.

File: pkgs/by-name/tl/tlock/package.nix [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+}:
+
+buildGoModule rec {
+  pname = "tlock";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "eklairs";
+    repo = "tlock";
+    rev = "v${version}";
```
[the remark this answers is not in the document]
Comment (SigmaSquadron) [reply pkgs/by-name/tl/tlock/package.nix:14]: This is non-standard syntax and has already been resolved here: https://github.com/NixOS/nixpkgs/pull/344896#discussion_r1778940895

[the remark this answers is not in the document]
Comment (FliegendeWurst) [reply pkgs/by-name/tl/tlock/package.nix:14]: refs/tags is the recommended approach. Otherwise the fetch fails if there is a branch with the same name.

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**